README.md
# GeoDescriberAI

This repository contains a [FastAPI](https://fastapi.tiangolo.com/) API and [Streamlit](https://streamlit.io/) APP that generates a description of a region based on a bounding box and a description request. 
It uses the [Overpass API](https://wiki.openstreetmap.org/wiki/Overpass_API) and the [OpenAI API](https://openai.com/blog/openai-api) for getting contextual data and geodescribing the region, respectively.

![](images/demo.png)

## Installation

1. Clone the repository: `git@github.com:Vizzuality/GeoDescriberAI.git`
2. Change into the repository directory: `cd GeoDescriberAI`
3. Create a virtual environment: `python -m venv env`
4. Activate the virtual environment: `source env/bin/activate`
5. Install the required packages: `pip install -r requirements.txt`
6. Create a `.env` file with your OpenAI API key:

    ```
    OPENAI_API_KEY=your_openai_api_key
    ```

## Usage

### FastAPI

This is a FastAPI-based API that generates a description of a region based on a bounding box (bbox) and a description request (chat). 
It utilizes external APIs for obtaining context data and generating descriptions.

1. To start the FastAPI server, use the following command:

```
uvicorn main:app --reload
```
 
2. Once the server is running, you can send HTTP requests to the API endpoints.

   **`/context` Endpoint**

   - URL: http://localhost:8000/context
   - Method: POST
   - Request Body: JSON object with the following properties:
     - `min_lon`: Minimum longitude of the bounding box (optional)
     - `min_lat`: Minimum latitude of the bounding box (optional)
     - `max_lon`: Maximum longitude of the bounding box (optional)
     - `max_lat`: Maximum latitude of the bounding box (optional)

   Example Request:

      ```commandline
      curl -X POST -H "Content-Type: application/json" -d '{
        "min_lon": -16.974,
        "min_lat": 27.986,
        "max_lon": -16.101,
        "max_lat": 28.595
      }' http://localhost:8000/context
      ```
   
   **`/description` Endpoint**

   - URL: http://localhost:8000/context
   - Method: POST
   - Request Body: JSON object with the following properties:
     - `bbox`: Bounding box object with the same properties as the `/context` endpoint
     - `chat`: Chat object with a `text` property containing the chat text

   Example Request:

      ```commandline
      curl -X POST -H "Content-Type: application/json" -d '{
        "bbox": {
          "min_lon": -16.974,
          "min_lat": 27.986,
          "max_lon": -16.101,  
          "max_lat": 28.595
        },
        "chat": {
          "text": "Describe the history, climate, and landscape of the region."
        }
      }' http://localhost:8000/description
      ```
   
   Response:

      ```
      {
         "description": "Welcome to the region of **Tenerife**, a beautiful island located in the Atlantic Ocean and part of the Canary Islands. The history of Tenerife dates back to the 15th century when the island was conquered by Spain. The island's indigenous people, known as Guanches, were the first to inhabit the island before the Spanish conquest. Today, the island is a popular tourist destination, attracting millions of visitors every year.\n\nThe climate of Tenerife is subtropical, with mild temperatures throughout the year. The island has a warm and dry climate, with temperatures ranging from 20°C to 30°C in the summer months and 15°C to 25°C in the winter months. The island's climate is influenced by the trade winds, which bring cool breezes from the Atlantic Ocean.\n\nThe landscape of Tenerife is diverse and unique, with a mix of volcanic and coastal features. The island is home to one of the world's largest volcanoes, **Teide**, which is also a UNESCO World Heritage Site. The volcanic landscape of Tenerife is characterized by steep cliffs, deep gorges, and rugged terrain. The island also has a beautiful coastline, with sandy beaches and crystal-clear waters. Tenerife is also home to two protected areas, the **Parque Rural de Teno** and the **Parque Rural de Anaga**, which are rich in biodiversity and offer stunning views of the island's landscape.\n\nIn summary, Tenerife is a region with a rich history, a subtropical climate, and a diverse landscape that includes volcanic mountains, rugged terrain, and beautiful coastlines."
      }
      ```

### Streamlit APP

After starting the FastAPI server run the Streamlit app, use the following command:

```
streamlit run app.py
```

This will launch the app in your default web browser.

1. Click the black square on the map
2. Draw a rectangle on the map
3. Provide your description request in the textbox below
4. Click on `Submit`
5. Wait for the computation to finish
   
## License
This project is licensed under the [MIT License](LICENSE).


CHANGELOG.md
# Changelog

## [Unreleased]

### Roadmap

- **Persistence of API Performance Metrics:** Implement a system to persist and analyze API performance over time.
- **Tests:** Add backend tests
- **CI/CD Pipelines:** Establish continuous integration and continuous deployment to streamline development and release
  processes.
- **Training and Deployment of an Open Source LLM:** Investigate and possibly include the training and deployment of an
  open-source Large Language Model within the project scope.
- **CT Pipeline:** It would be hardcore-level cool to implement a continuous training pipeline.
- **Answers correctness:** Add ability to flag an answer as correct or incorrect.
- **Abstract the data access layer:** Make it extensible to work with other databases / DbaS
- **Do you have cool ideas?** Please share them!

## [Version 0.0.1]

### Added

#### Infrastructure Setup

- **Backend:** FastAPI framework used for creating the API.
- **Frontend:** Streamlit application for user interactions.
- **Database:** MongoDB for data persistence.
- **Cloud Providers:** DO integration for hosting and related services.

#### Q&A Persistence Implementation

- Added persistence for questions and answers, enabling historical tracking, analytics and hopefully data to fine-tune
  the thing.

#### Authentication System

- Auth system to avoid bots and human trolls.

---

This changelog follows the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format, and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


main.py
from typing import Annotated

from fastapi import FastAPI, Depends, HTTPException
from fastapi.security import OAuth2PasswordRequestForm
from pydantic import BaseModel

from auth.auth import create_token, get_current_user
from database.repository import get_mongo_repository
from processing.context_data import get_overpass_api_response
from processing.description import get_openai_api_response
from questions_answers.repository import get_qa_repository, QuestionAnswerRepository
from users.repository import get_user_repository, UserRepository

app = FastAPI()
mongo_repository = get_mongo_repository()


class Bbox(BaseModel):
    min_lon: float = -16.974
    min_lat: float = 27.986
    max_lon: float = -16.101
    max_lat: float = 28.595


class Chat(BaseModel):
    text: str = "Describe the history, climate, and landscape of the region."


@app.on_event("startup")
async def startup_event():
    print("Starting up...")
    is_connected = await mongo_repository.test_connection()
    if is_connected:
        print("Connected to MongoDB")
    else:
        print("Failed to connect to MongoDB")


@app.post("/login")
async def login(form_data: Annotated[OAuth2PasswordRequestForm, Depends()],
                user_repo: UserRepository = Depends(get_user_repository, use_cache=True)):
    username, password = form_data.username, form_data.password
    user = await user_repo.get_user_by_credentials(username, password)
    if user:
        token = create_token(username)
        return {"access_token": token}
    else:
        raise HTTPException(status_code=401, detail="Invalid credentials")


@app.post("/context")
async def get_context_data(bbox: Bbox):
    context_data = await get_overpass_api_response(bbox)
    return context_data


@app.post("/description", tags=["Protected"])
async def get_description(bbox: Bbox, chat: Chat,
                          repo: QuestionAnswerRepository = Depends(get_qa_repository, use_cache=True),
                          current_user: str = Depends(get_current_user)):
    context_data = await get_overpass_api_response(bbox)
    description = await get_openai_api_response(data=context_data, chat=chat)
    await repo.save_question_answer(question=chat.text, answer=description)
    return {"description": description}
